File: statistics/1_EDA_dis/1_EDA_dis.py
# -*- coding: utf-8 -*-

import os
os.chdir(r'D:\cs\data_ml_f\1\1\1cs\data\0\data')

import numpy as np
import pandas as pd


state = pd.read_csv('state.csv')

# <Estimation of Location>

# 1. mean group

# (1)arithmetic mean
# > two ways
state['Population'].mean()

np.mean(state['Population'])
# > handling missing values
np.nanmean(state['Population'])

# (2)geometric mean
# >positive number

# use package scipy.stats
import scipy.stats as stats
stats.gmean(state['Population'])

# (3) trimmed-mean
# >trim=0.1 for each end
stats.trim_mean(state['Population'], 0.1)

# weighted-mean
np.average(state['Murder.Rate'], weights=state['Population'])

# 2. median group

# (1) median
# > two ways
state['Population'].median()

np.median(state['Population'])

# (2) weighted-median

# use package wquantiles
import wquantiles
wquantiles.median(state['Murder.Rate'], weights=state['Population'])

# 3. mode
# > generally not used for numeric data
stats.mode(state['Population'])

#==============================================================================#

# <Estimation of Variability>

# 1. deviation group

# (1) mean absolute deviation
state["Population"].mad()

# (2) variance
# > two ways
# > in pandas, the denominator is n-1
state['Population'].var()

# > in np, the default denominator is n
# > but we can change it
np.var(state['Population'], ddof=1)

# (3) std
# > in Python, it is std,;in R it is sd
# > two ways
# > in pandas, the denominator is n-1
state['Population'].std()

# > in np, the default denominator is n
# > but we can change it
np.std(state['Population'], ddof=1)

# (4) MAD -- median absolute deviation
# > use package statsmodels.robust.scale
from statsmodels.robust.scale import mad
mad(state['Population'])

# 2. order statistics

# (1) range
# > two ways
np.ptp(state['Population'])

state['Population'].max() - state['Population'].min()


# (2) percentile or quantile
state['Population'].quantile(0.5)
# multiple quantile
state['Population'].quantile([0.75, 0.25])

# > np way
np.quantile(state['Population'], q=[0.75, 0.25])

# (3) IQR .75 quantile - .25 quantile
# > two ways
stats.iqr(state['Population'])

state['Population'].quantile(0.75) - state['Population'].quantile(0.25)

# 3. SE, SEM
# define function
def se(df, na=False):
    if na:
        df = df[df.notnull()]
    from math import sqrt
    return sqrt(df.var() / len(df))

se(state['Population'])

# or use the package stats
stats.sem(state['Population'])

#==============================================================================#

# <Estimation of Shape>

# 1. Skewness
# > two ways
state['Population'].skew()

# bias=False --  the calculations are corrected for statistical bias
stats.skew(state['Population'], bias=False)

# 2. Kurtosis
# > two ways
state['Population'].kurtosis()

stats.kurtosis(state['Population'], bias=False)


#==============================================================================#
#==============================================================================#


# Distribution
# there are so many dis. we just can't show them all
# So, we use normal dis. and t dis. to do the presentation

# 1. Normal
norm_dis = stats.norm()
# > loc -- mean, scale -- std
norm_dis_1 = stats.norm(loc=2, scale=3)

# (1) density
norm_dis_1.pdf(5)

# (2) CDF
norm_dis_1.cdf(1)

# (3) sf = 1 - CDF
norm_dis_1.sf(1)

# (4) ppf - inverse of CDF
norm_dis.ppf(0.975)
norm_dis_1.ppf(norm_dis_1.cdf(1))

# (5) ISF - inverse of SF
# > get the 1 - probability, use ppf to get the value
norm_dis.isf(0.025)

# (6) RVS
norm_dis.rvs(10)

# 2. t-dis.
t_dis = stats.t(df=5)

# (1) density
t_dis.pdf(t_dis.ppf(0.95))

# (2) CDF
t_dis.cdf(t_dis.ppf(0.95))

# (3) sf = 1 - CDF
t_dis.sf(t_dis.ppf(0.95))

# (4) ppf - inverse of CDF
t_dis.ppf(0.95)

# (5) ISF - inverse of SF
# > get the 1 - probability, use ppf to get the value
t_dis.isf(0.05)

# (6) RVS
t_dis.rvs(10)

File: statistics/1_EDA_dis/1_EDA_dis.r
state <- read.csv('D:/cs/data_ml_f/1/1/1cs/data/0/data/state.csv')
# <Estimation of Location>

# 1. mean group

# (1)arithmetic mean
mean(state$Population)

# (2)geometric mean
# >positive number

# use package psych
install.packages("psych")
library(psych)

geometric.mean(state$Population)

# (3) trimmed-mean
# >trim=0.1 for each end
mean(state$Population, trim=0.1)

# weighted-mean
weighted.mean(state$Murder.Rate, w=state$Population)

# 2. median group

# (1) median
median(state$Population)

# (2) weighted-median

# use matrixStats package
install.packages("matrixStats")
library(matrixStats)
weightedMedian(state$Murder.Rate, w=state$Population)

# 3. mode
# there are no built-in function for mode in R

#==============================================================================#
#==============================================================================#

# <Estimation of Variability>

# 1. deviation group

# (1) mean absolute deviation
# use package ie2misc
# > https://search.r-project.org/CRAN/refmans/ie2misc/html/madstat.html
install.packages("ie2misc")
library(ie2misc)

madstat(state$Population)

# (2) variance
# > the denominator is n-1
var(state$Population)

# (3) std
# > the denominator is n-1
# > in Python, it is std, here it is sd
sd(state$Population)

# (4) MAD -- median absolute deviation
mad(state$Population)

# 2. order statistics

# (1) range
range(state$Population)
# if you want the length of the data set
# there are way ways
range(state$Population)[2] - range(state$Population)[1]
max(state$Population) - min(state$Population)

# (2) percentile or quantile
quantile(state$Population, .5)
# multiple quantile
quantile(state$Population, p=c(.75, .25))

# (3) IQR .75 quantile - .25 quantile
IQR(state$Population)

# 3. SE, SEM
# define function
se <- function(x, na.rm = FALSE) {
  if (na.rm) x <- na.omit(x)
  sqrt(var(x) / length(x))
}

se(state$Population)

# or use the package plotrix
install.packages("plotrix")
plotrix::std.error(state$Population)


#==============================================================================#
#==============================================================================#

# <Estimation of Shape>

# 1. Skewness
# use package moments
install.packages("moments")
# > it is the same as  "biased version" in Python
moments::skewness(state$Population)

# 2. Kurtosis
# use package moments
# > it is the same as  "biased version" plus 3 in Python
moments::kurtosis(state$Population)


#==============================================================================#
#==============================================================================#

# Distribution
# there are so many dis. we just can't show them all
# So, we use binomial dis., poisson dis. and normal dis. to do the presentation

# 1. Binomial
# (1) PMF -- x for success num.
dbinom(x=2, size=5, p=0.1)

# (2) CDF
pbinom(2, 5, 0.1)

# (3) generate random numbers from binomial dis.
# > n -- the size you want
# > the returned values are the nums of successes
rbinom(n=10, size=100, p=0.1)

# 2. Poisson
# (1) PMF
dpois(x=3, lambda=2)

# (2) CDF
ppois(6, 6)

# (3) random variates
rpois(n=5, 10)

# 3. Normal
# (1) ISF
qnorm(0.975, 0, 1)

#==============================================================================#

# <functions>
# 1. Normal

# (1) density
dnorm(5 ,mean=5, sd=1)

# (2) CDF
pnorm(1, mean=2, sd=3)

# (3) sf = 1 - CDF
1 - pnorm(1, mean=2, sd=1)

# (4) ppf - inverse of CDF
qnorm(pnorm(1, mean=2, sd=3), mean=2, sd=3)

# (5) ISF - inverse of SF
# > get the 1 - probability, use ppf to get the value
qnorm(1-0.95, mean=0, sd=1)

# (6) RVS
rnorm(5, mean=2, sd=3)

# 2. t-dis.

# (1) density
dt(qt(0.95, 5), 5)

# (2) CDF
pt(qt(0.95, 5), 5)

# (3) sf = 1 - CDF
1 - pt(qt(0.95, 5), 5)

# (4) ppf - inverse of CDF
qt(0.95, 5)

# (5) ISF - inverse of SF
# > get the 1 - probability, use ppf to get the value
qt(1-0.05, 5)

# (6) RVS
rt(5, 5)
